File: noesis-mcp/src/validateFrontmatter.ts
import fs from "node:fs";
import path from "node:path";

// Porta de scripts/validate_frontmatter.py pra dentro do noesis-mcp — mesma
// lógica, mesmas regras (RULES/REQUIRED_COMMON), reescrita em TS pra rodar
// em processo. Motivo: create_note/create_relation (vault.ts#runValidator)
// shelavam pra `python3 scripts/validate_frontmatter.py`, o que quebra em
// qualquer máquina sem Python 3 instalado — inviável pro Aurora Desktop
// empacotado, que precisa funcionar pra usuário não-técnico sem instalar
// nada além do instalador do app (ver decisions/ADR-0008-vault-por-instalacao.md).
// O script Python continua existindo e sendo a validação usada pelo hook de
// pre-commit em `scripts/pre-commit.sh` (workflow de quem edita o vault
// direto via git) — as duas cópias das regras precisam ficar em sincronia
// manualmente; mudou uma, muda a outra.
//
// Parser de frontmatter é deliberadamente mínimo (só `key: value` de
// primeiro nível, sem listas/objetos aninhados) — mesma limitação
// documentada no script original.

type Frontmatter = Record<string, string>;

const REQUIRED_COMMON = ["id", "type", "status", "created"];

interface TypeRule {
  extra: string[];
  allowed: Record<string, Set<string>>;
}

const RULES: Record<string, TypeRule> = {
  goal: {
    extra: ["horizon", "origin", "confidence", "progress", "success_criteria", "review_cycle"],
    allowed: {
      horizon: new Set(["short", "mid", "long"]),
      status: new Set(["active", "paused", "achieved", "killed"]),
      origin: new Set(["declared"]),
    },
  },
  habit: {
    extra: ["direction", "origin", "trigger", "frequency_target"],
    allowed: {
      direction: new Set(["build", "extinguish"]),
      status: new Set(["active", "consolidated", "abandoned"]),
      origin: new Set(["declared"]),
    },
  },
  hypothesis: {
    extra: ["confidence", "origin"],
    allowed: { origin: new Set(["inferred", "aurora", "declared", "co-created", "daemon", "session"]) },
  },
  project: { extra: ["domain"], allowed: { status: new Set(["active", "paused", "done"]) } },
  decision: { extra: [], allowed: {} },
  identity: { extra: ["version", "confidence", "mutable_by_system"], allowed: {} },
  foundation: { extra: ["version", "confidence"], allowed: {} },
  value: { extra: ["origin"], allowed: { origin: new Set(["declared"]) } },
  skill: { extra: [], allowed: {} },
  conflict: { extra: [], allowed: {} },
  meta: { extra: [], allowed: {} },
};

const ID_MUST_MATCH_FILENAME = new Set(["goal", "habit", "project", "value", "skill"]);

function stripChar(s: string, ch: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && s[start] === ch) start++;
  while (end > start && s[end - 1] === ch) end--;
  return s.slice(start, end);
}

// Python: v.strip('"').strip("'") — strip() remove ocorrências repetidas
// nas duas pontas, um caractere por vez, em duas passadas sequenciais.
function stripQuotes(v: string): string {
  return stripChar(stripChar(v, '"'), "'");
}

function parseFrontmatter(text: string): Frontmatter | null {
  const m = text.match(/^---\s*\n([\s\S]*?)\n---\s*\n/);
  if (!m) return null;
  const fm: Frontmatter = {};
  for (const raw of m[1].split("\n")) {
    const trimmed = raw.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const indent = raw.length - raw.trimStart().length;
    const colonIdx = trimmed.indexOf(":");
    if (indent === 0 && colonIdx !== -1) {
      const key = trimmed.slice(0, colonIdx).trim();
      let value = trimmed.slice(colonIdx + 1);
      value = value.split("#")[0].trim();
      value = stripQuotes(value);
      fm[key] = value;
    }
  }
  return fm;
}

// Python: datetime.date.fromisoformat(c[:10]) — exige YYYY-MM-DD estrito
// (mês 01-12, dia válido pro mês/ano, incluindo ano bissexto).
function isIsoDate(s: string): boolean {
  const m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(s.slice(0, 10));
  if (!m) return false;
  const year = Number(m[1]);
  const month = Number(m[2]);
  const day = Number(m[3]);
  if (month < 1 || month > 12) return false;
  const daysInMonth = new Date(year, month, 0).getDate();
  return day >= 1 && day <= daysInMonth;
}

function checkFile(absPath: string): string[] {
  const errs: string[] = [];
  const text = fs.readFileSync(absPath, "utf-8");
  const fm = parseFrontmatter(text);
  if (fm === null) {
    return [`${absPath}: sem frontmatter YAML delimitado por ---`];
  }

  for (const f of REQUIRED_COMMON) {
    if (!fm[f]) errs.push(`${absPath}: campo obrigatório ausente: ${f}`);
  }

  const t = fm.type ?? "";
  const rule = RULES[t];
  if (!rule) {
    errs.push(`${absPath}: type desconhecido: '${t}' (permitidos: ${Object.keys(RULES).sort().join(", ")})`);
  } else {
    for (const f of rule.extra) {
      if (!fm[f]) errs.push(`${absPath}: [${t}] campo obrigatório ausente: ${f}`);
    }
    for (const [field, vals] of Object.entries(rule.allowed)) {
      const v = fm[field] ?? "";
      if (v && !vals.has(v)) {
        errs.push(`${absPath}: [${t}] ${field}='${v}' inválido (permitidos: ${[...vals].sort().join(", ")})`);
      }
    }
  }

  const fid = fm.id ?? "";
  if (fid && ID_MUST_MATCH_FILENAME.has(t)) {
    const stem = path.basename(absPath, ".md");
    if (stem !== fid) {
      errs.push(`${absPath}: id '${fid}' difere do nome do arquivo '${stem}'`);
    }
  }

  for (const numField of ["confidence", "progress"]) {
    const v = fm[numField];
    if (v) {
      const x = Number(v);
      if (Number.isNaN(x)) {
        errs.push(`${absPath}: ${numField}='${v}' não é numérico`);
      } else if (x < 0 || x > 1) {
        errs.push(`${absPath}: ${numField}=${v} fora de [0,1]`);
      }
    }
  }

  const created = fm.created ?? "";
  if (created && !isIsoDate(created)) {
    errs.push(`${absPath}: created='${created}' não é data ISO (YYYY-MM-DD)`);
  }

  return errs;
}

export interface ValidationResult {
  ok: boolean;
  output: string;
}

/** Mesmas regras de scripts/validate_frontmatter.py, em processo — sem spawnar python3. */
export function validateFrontmatterFiles(absPaths: string[]): ValidationResult {
  const allErrs: string[] = [];
  for (const absPath of absPaths) {
    if (!absPath.endsWith(".md") || !fs.existsSync(absPath)) continue;
    allErrs.push(...checkFile(absPath));
  }
  if (allErrs.length > 0) {
    return { ok: false, output: "VALIDAÇÃO FALHOU:\n" + allErrs.map((e) => `  ✗ ${e}`).join("\n") };
  }
  return { ok: true, output: `✓ ${absPaths.length} nota(s) válida(s). Frontmatter OK.` };
}

File: scripts/validate_frontmatter.py
#!/usr/bin/env python3
"""
NOESIS — Validador de frontmatter (Fase 0 do roadmap)
Uso:
  python3 scripts/validate_frontmatter.py            # valida o vault inteiro
  python3 scripts/validate_frontmatter.py file.md    # valida arquivos específicos
Como pre-commit (na raiz do repo):
  ln -s ../../scripts/pre-commit.sh .git/hooks/pre-commit
Sem dependências externas (não requer PyYAML): parser mínimo próprio.
"""
import re, sys, datetime
from pathlib import Path

VAULT = Path(__file__).resolve().parent.parent

REQUIRED_COMMON = ["id", "type", "status", "created"]
RULES = {
    # type: (campos obrigatórios extras, valores permitidos por campo)
    "goal":       (["horizon", "origin", "confidence", "progress", "success_criteria", "review_cycle"],
                   {"horizon": {"short", "mid", "long"},
                    "status": {"active", "paused", "achieved", "killed"},
                    "origin": {"declared"}}),
    "habit":      (["direction", "origin", "trigger", "frequency_target"],
                   {"direction": {"build", "extinguish"},
                    "status": {"active", "consolidated", "abandoned"},
                    "origin": {"declared"}}),
    "hypothesis": (["confidence", "origin"],
                   {"origin": {"inferred", "aurora", "declared", "co-created", "daemon", "session"}}),
    "project":    (["domain"], {"status": {"active", "paused", "done"}}),
    "decision":   ([], {}),
    "identity":   (["version", "confidence", "mutable_by_system"], {}),
    "foundation": (["version", "confidence"], {}),
    "value":      (["origin"], {"origin": {"declared"}}),
    "skill":      ([], {}),
    "conflict":   ([], {}),
    "meta":       ([], {}),
}

IGNORED_DIRS = {".git", ".obsidian", "node_modules", "scripts", "events", "research"}
IGNORED_FILES = {"SETUP.md", "README.md", "CLAUDE.md", "00-INDEX.md"}

def parse_frontmatter(text):
    m = re.match(r"^---\s*\n(.*?)\n---\s*\n", text, re.DOTALL)
    if not m:
        return None
    fm, stack = {}, []
    for raw in m.group(1).splitlines():
        if not raw.strip() or raw.strip().startswith("#"):
            continue
        indent = len(raw) - len(raw.lstrip())
        line = raw.strip()
        if indent == 0 and ":" in line:
            k, _, v = line.partition(":")
            v = v.split("#")[0].strip().strip('"').strip("'")
            fm[k.strip()] = v if v else {}
    return fm

def check_file(path):
    errs = []
    text = path.read_text(encoding="utf-8", errors="replace")
    fm = parse_frontmatter(text)
    if fm is None:
        return [f"{path}: sem frontmatter YAML delimitado por ---"]
    for f in REQUIRED_COMMON:
        if f not in fm or fm[f] in ("", {}):
            errs.append(f"{path}: campo obrigatório ausente: {f}")
    t = fm.get("type", "")
    if t not in RULES:
        errs.append(f"{path}: type desconhecido: '{t}' (permitidos: {', '.join(sorted(RULES))})")
    else:
        extra, allowed = RULES[t]
        for f in extra:
            if f not in fm or fm[f] in ("", {}):
                errs.append(f"{path}: [{t}] campo obrigatório ausente: {f}")
        for field, vals in allowed.items():
            v = fm.get(field, "")
            if isinstance(v, str) and v and v not in vals:
                errs.append(f"{path}: [{t}] {field}='{v}' inválido (permitidos: {', '.join(sorted(vals))})")
    # id deve bater com o nome do arquivo para tipos de entidade
    fid = fm.get("id", "")
    if isinstance(fid, str) and fid and t in {"goal", "habit", "project", "value", "skill"}:
        if path.stem != fid:
            errs.append(f"{path}: id '{fid}' difere do nome do arquivo '{path.stem}'")
    # confidence e progress no intervalo [0,1]
    for numf in ("confidence", "progress"):
        v = fm.get(numf, "")
        if isinstance(v, str) and v:
            try:
                x = float(v)
                if not (0.0 <= x <= 1.0):
                    errs.append(f"{path}: {numf}={v} fora de [0,1]")
            except ValueError:
                errs.append(f"{path}: {numf}='{v}' não é numérico")
    # importance (opcional, ADR-0010): quando presente, numérico em [0,10]
    v = fm.get("importance", "")
    if isinstance(v, str) and v:
        try:
            x = float(v)
            if not (0.0 <= x <= 10.0):
                errs.append(f"{path}: importance={v} fora de [0,10]")
        except ValueError:
            errs.append(f"{path}: importance='{v}' não é numérico")
    # created deve ser data ISO
    c = fm.get("created", "")
    if isinstance(c, str) and c:
        try:
            datetime.date.fromisoformat(c[:10])
        except ValueError:
            errs.append(f"{path}: created='{c}' não é data ISO (YYYY-MM-DD)")
    return errs

def main(argv):
    raw_targets = [Path(a) for a in argv[1:]] if len(argv) > 1 else list(VAULT.rglob("*.md"))
    # Correcao 2026-07-17: o filtro de IGNORED_DIRS/IGNORED_FILES so era
    # aplicado no modo "vault inteiro". Isso fazia o pre-commit hook (que
    # sempre passa argv explicito com os arquivos staged) falhar em
    # arquivos como SETUP.md, que sao intencionalmente isentos de
    # frontmatter. Aplicamos o mesmo filtro nos dois modos agora.
    targets = [
        p for p in raw_targets
        if not any(part in IGNORED_DIRS for part in p.parts) and p.name not in IGNORED_FILES
    ]
    all_errs = []
    for p in targets:
        if p.suffix == ".md" and p.exists():
            all_errs += check_file(p)
    if all_errs:
        print("VALIDAÇÃO FALHOU:\n" + "\n".join(f"  ✗ {e}" for e in all_errs))
        return 1
    print(f"✓ {len(targets)} nota(s) válida(s). Frontmatter OK.")
    return 0

if __name__ == "__main__":
    sys.exit(main(sys.argv))
